Sources/Basics/HTTPClient.swift

```diff
@@ -0,0 +1,262 @@
+        case delete
+    }
+
+    public struct Options {
+        var addUserAgent: Bool
+        var validResponseCodes: [Int]?
+        var timeout: DispatchTimeInterval?
+        var cachePolicy: NSURLRequest.CachePolicy?
+        var callbackQueue: DispatchQueue?
+
+        public init(addUserAgent: Bool = true,
+                    validResponseCodes: [Int]? = nil,
+                    timeout: DispatchTimeInterval? = nil,
+                    cachePolicy: NSURLRequest.CachePolicy? = nil,
+                    callbackQueue: DispatchQueue? = nil) {
+            self.addUserAgent = addUserAgent
+            self.validResponseCodes = validResponseCodes
+            self.timeout = timeout
+            self.cachePolicy = cachePolicy
+            self.callbackQueue = callbackQueue
+        }
+    }
+}
+
+public struct HTTPClientResponse {
+    let statusCode: Int
+    let statusText: String?
+    let headers: HTTPClientHeaders
+    let body: Data?
+
+    public init(statusCode: Int,
+                statusText: String? = nil,
+                headers: HTTPClientHeaders = .init(),
+                body: Data? = nil) {
+        self.statusCode = statusCode
+        self.statusText = statusText
+        self.headers = headers
+        self.body = body
+    }
+}
+
+public struct HTTPClientHeaders: Sequence {
+    // TODO: optimize
+    private var headers: [Item]
+
+    public init(_ headers: [Item] = []) {
+        self.headers = headers
+    }
+
+    // TODO: optimize
+    public func contains(_ name: String) -> Bool {
+        self.headers.contains(where: { $0.name.lowercased() == name.lowercased() })
+    }
+
+    public mutating func add(name: String, value: String) {
+        self.add(Item(name: name, value: value))
+    }
+
+    public mutating func add(_ item: Item) {
+        self.headers.append(item)
+    }
+
+    // TODO: optimize
+    public func get(name: String) -> [String] {
+        self.headers.filter { $0.name.lowercased() == name.lowercased() }.map { $0.value }
+    }
+
+    public func makeIterator() -> IndexingIterator<[Item]> {
+        self.headers.makeIterator()
+    }
+
+    public typealias Item = (name: String, value: String)
+}
+
+// MARK: - SPMHTTPClient
+
+public struct SPMHTTPClient: HTTPClient {
+    private let configuration: Configuration
+
+    public init(configuration: Configuration) {
+        self.configuration = configuration
+    }
+
+    public func execute(request: Request, callback: @escaping (Result<Response, Error>) -> Void) {
+        let session = URLSession(configuration: self.configuration.urlSessionConfiguration())
+
+        var request = request
+        if let additionalHeaders = self.configuration.requestHeaders {
+            additionalHeaders.forEach { header in
+                request.headers.add(header)
+            }
+        }
+        if request.options.addUserAgent, !request.headers.contains("User-Agent") {
+            // TODO: need SwiftPM version here, how can we get it?
```

Comment:
Is it `Versioning.currentVersion.completeDisplayString`?

Reply:
Yep